### PR TITLE
Fixes a ton of mirror bugs, makes magic mirrors more open-minded

### PR DESCRIPTION
**Pull request by ATH1909**
## About The Pull Request

Magic mirrors (including pride mirrors) will no longer pull up the normal mirror interface before their own when interacted with.

You can no longer avoid the curse of a pride mirror with some species by just walking away after finishing some, but not all of your changes to one of your features (i.e. walking away after choosing a species but before choosing your skin tone).

For magic mirrors, species selection has been moved out of the "race" feature category and into its own "species" category.

Magic mirrors can now modify your gender independently of your sex, and you can become nonbinary using one if you so desire.

Characters with a female body type can select a new facial hair style at any mirror, just like characters with male body types can, and are no longer forced to shave their facial hair when they use a mirror. Note that, according to mothblocks, characters with female body types are already capable of giving themselves beards during character creation, and mirrors were just never updated to reflect this.

Mirror code has been brought closer to our standards.

Some mirror-related descriptions have been tweaked. Nothing to see here.

Probably fixes half of https://github.com/tgstation/tgstation/issues/62691.

Testing for this PR will come Soon(TM), after Manuel's peak hours have passed.

## Why It's Good For The Game

Needing to dismiss the normal mirror interface to use a magic mirror is really annoying and also probably misleading for new players.

Our current mirror code is... Okay, I think the best way to summarize our current mirror code is that it uses "eyes" in the list of features you can opt to change, then checks for BODY_ZONE_PRECISE_EYES in its big switch statement. If the definition of BODY_ZONE_PRECISE_EYES ever changes (without this PR being merged), magic mirrors will silently lose their ability to change your eye color.

Also, like, the gender-related changes are probably nice to have.

image

## Changelog

:cl: ATHATH
fix: Magic mirrors (including pride mirrors) will no longer pull up the normal mirror interface before their own when interacted with.
fix: You can no longer avoid the curse of a pride mirror with some species by just walking away after finishing some, but not all of your changes to one of your features (i.e. walking away after choosing a species but before choosing your skin tone).
fix: For magic mirrors, species selection has been moved out of the "race" feature category and into its own "species" category.
fix: Magic mirrors can now modify your gender independently of your sex, and you can become nonbinary using one if you so desire.
fix: Characters with a female body type can select a new facial hair style at any mirror, just like characters with male body types can, and are no longer forced to shave their facial hair when they use a mirror. Note that, according to mothblocks, characters with female body types are already capable of giving themselves beards during character creation, and mirrors were just never updated to reflect this.
code: Mirror code has been brought closer to our standards.
spellcheck: Some mirror-related descriptions have been tweaked. Nothing to see here.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/mirror.dm [hunk 1]
```diff
@@ -23,32 +36,39 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
+	mirror_stuff(user)
+
+//Mirror man, mirror man, does whatever a mirror can.
+/obj/structure/mirror/mirror_stuff(mob/living/carbon/human/stylist)
+	//see code/modules/mob/dead/new_player/preferences.dm at approx line 545 for comments!
+	//this is largely copypasted from there.
+
+	//handle normal hair
+	if(!HAS_TRAIT(stylist, TRAIT_BALD))
+		var/new_style = input(stylist, "Select a hairstyle", "Grooming")  as null|anything in GLOB.hairstyles_list
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:51]: `tgui_input_list()` is the tgui version of input, i'm gonna say it's a better idea to use that here

File: code/game/objects/structures/mirror.dm [hunk 2]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
+
+
+		if("eyes")
+			var/new_eye_color = input(stylist, "What is the color of our eyes again?", "Eye Color",stylist.eye_color) as color|null
+			if(!stylist.canUseTopic(src, BE_CLOSE, FALSE, NO_TK) || !new_eye_color)
+				return
+			stylist.eye_color = sanitize_hexcolor(new_eye_color)
+			stylist.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+			stylist.update_body()
+			curse(stylist)
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:303]: shouldn't you be checking curse's return here?

File: code/game/objects/structures/mirror.dm [hunk 3]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
-	if(choice)
-		curse(user)
-
-/obj/structure/mirror/magic/proc/curse(mob/living/user)
-	return
+
+			stylist.facial_hair_color = sanitize_hexcolor(new_face_color)
+			stylist.dna.update_ui_block(DNA_FACIAL_HAIR_COLOR_BLOCK)
+			stylist.update_hair()
+			curse(stylist)
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:292]: and here?

File: code/game/objects/structures/mirror.dm [hunk 4]
```diff
@@ -23,32 +36,39 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
-			H.hairstyle = new_style
+		else if(new_style)
+			stylist.hairstyle = new_style
+			stylist.update_hair()
+			if(curse(stylist))
+				return
 
-		H.update_hair()
+	//handle facial hair
+	var/new_facial_hair_style = input(stylist, "Select a facial hairstyle", "Grooming")  as null|anything in GLOB.facial_hairstyles_list
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:61]: `tgui_input_list()` works here too!

File: code/game/objects/structures/mirror.dm [hunk 5]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
 	return ..()
 
-/obj/structure/mirror/magic/attack_hand(mob/user, list/modifiers)
-	. = ..()
-	if(.)
-		return
-	if(!ishuman(user))
-		return
+/obj/structure/mirror/magic/mirror_stuff(mob/living/carbon/human/stylist)
+	var/choice = input(stylist, "Something to change?", "Magical Grooming") as null|anything in list("name", "species", "color", "gender", "sex", "hairstyle", "hair color", "eyes")
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:167]: `tgui_input_list()` works here too!

File: code/game/objects/structures/mirror.dm [hunk 6]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
+			stylist.name = newname
+			if(stylist.dna)
+				stylist.dna.real_name = newname
+			if(stylist.mind)
+				stylist.mind.name = newname
+			curse(stylist)
+
+		if("species")
+			var/newspecies
+			var/specieschoice = input(stylist, "What species are we again?", "Species change") as null|anything in choosable_races
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:187]: `tgui_input_list()` works here too!

File: code/game/objects/structures/mirror.dm [hunk 7]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
+		if("color")
+			if(!stylist.dna.species.use_skintones && !(MUTCOLORS in stylist.dna.species.species_traits))
+				to_chat(stylist, span_notice("Your species doesn't have variant colorings."))
+				return
 
-			if(MUTCOLORS in H.dna.species.species_traits)
-				var/new_mutantcolor = input(user, "Choose your skin color:", "Race change",H.dna.features["mcolor"]) as color|null
-				if(!user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+			if(stylist.dna.species.use_skintones)
+				var/new_s_tone = input(stylist, "What is our skin tone again?", "Race change")  as null|anything in GLOB.skin_tones
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:206]: `tgui_input_list()` works here too!

File: code/game/objects/structures/mirror.dm [hunk 8]
```diff
@@ -143,129 +163,141 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
+					stylist.update_hair()
+					stylist.update_body_parts()
+					stylist.update_mutations_overlay()
+					curse(stylist)
+				else
+					to_chat(stylist, span_notice("Invalid color. Your color is not bright enough."))
 
 		if("gender")
-			if(!(H.gender in list("male", "female"))) //blame the patriarchy
+			var/attackhelicopter = input(stylist, "What do we identify as again?", "Gender change") as null|anything in list(MALE, FEMALE, PLURAL)
```
Comment (tralezab) [inline code/game/objects/structures/mirror.dm:238]: `tgui_input_list()` works here too!

**Review by tralezab — CHANGES_REQUESTED** [review]
cool stuff, lil' cursory review from me dawg

File: code/game/objects/structures/mirror.dm [hunk 9]
```diff
@@ -1,20 +1,33 @@
...
+/obj/structure/mirror/examine(mob/user)
+	. = ..()
+	if(desc != "It's you!") //I'm really not a fan of hardcoding this, but I don't see another way to do this
+		return
+	if(user.mind && user.mind.has_antag_datum(/datum/antagonist, TRUE) && user.key)
+		return "It's me, [user.key]." //uses the player's OOC name, not their IC one
+	if(SSshuttle.emergency && SSshuttle.emergency.mode == SHUTTLE_ENDGAME)
+		return "Still just you, [user.real_name]."
+	if(world.time >= creation_time + 60 MINUTES)
+		return "Despite everything, it's still you."
```
[the remark this answers is not in the document]
Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:30]: but that doesn't replace the description, it just adds onto the end of it

File: code/game/objects/structures/mirror.dm [hunk 10]
```diff
@@ -23,32 +36,39 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
-		if(!user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+	if(!ishuman(user))
+		return
+
+	mirror_stuff(user)
+
+//Mirror man, mirror man, does whatever a mirror can.
+/obj/structure/mirror/mirror_stuff(mob/living/carbon/human/stylist)
+	//see code/modules/mob/dead/new_player/preferences.dm at approx line 545 for comments!
+	//this is largely copypasted from there.
```
[the remark this answers is not in the document]
Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:47]: this wasn't my comment, it's in our current mirror code as well

File: code/game/objects/structures/mirror.dm (hunk 2 above)
Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:303]: nope, I only need to check its return if there's stuff after it

File: code/game/objects/structures/mirror.dm (hunk 9 above)
[the remark this answers is not in the document]
Comment (tralezab) [reply code/game/objects/structures/mirror.dm:30]: yeah but you're kinda throwing away a lot that examine should be sending though. Maybe you should set desc and then call `. = ..()`?

```dm
if(antagonist)
    desc = "its me the bad guy"
. = ..() //builds the text with the desc
desc = initial(desc)
```

File: code/game/objects/structures/mirror.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (tralezab) [reply code/game/objects/structures/mirror.dm:47]: you have successfully thought about whether it was appropriate to include, i'm happy >:)

File: code/game/objects/structures/mirror.dm (hunk 2 above)
Comment (tralezab) [reply code/game/objects/structures/mirror.dm:303]: *nod

File: code/game/objects/structures/mirror.dm (hunk 9 above)
[the remark this answers is not in the document]
Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:30]: but what if two people try to examine it at almost the same time?

[the remark this answers is not in the document]
Comment (Fikou) [reply code/game/objects/structures/mirror.dm:30]: no

File: code/game/objects/structures/mirror.dm [hunk 11]
```diff
@@ -23,32 +36,39 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/mirror, 28)
...
-		else
-			H.facial_hairstyle = "Shaved"
-
-		//handle normal hair
-		var/new_style = input(user, "Select a hairstyle", "Grooming")  as null|anything in GLOB.hairstyles_list
-		if(!user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+	if(!ishuman(user))
+		return
+
+	mirror_stuff(user)
```
[the remark this answers is not in the document]
Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:42]: 1984

File: code/game/objects/structures/mirror.dm (hunk 9 above)
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/game/objects/structures/mirror.dm:30]: byond isnt multi-threaded, and none of this yields

File: code/game/objects/structures/mirror.dm [hunk 12]
```diff
@@ -1,20 +1,33 @@
...
 
 /obj/structure/mirror/Initialize(mapload)
 	. = ..()
 	if(icon_state == "mirror_broke" && !broken)
 		atom_break(null, mapload)
+	creation_time = world.time
+
+/obj/structure/mirror/examine(mob/user)
+	. = ..()
+	if(desc != "It's you!") //I'm really not a fan of hardcoding this, but I don't see another way to do this
```
Comment (Mothblocks) [inline code/game/objects/structures/mirror.dm:23]:
```suggestion
	if(desc != initial(desc))
```

Comment (ATH1909) [reply code/game/objects/structures/mirror.dm:23]: I don't want this to apply to the mirror subtypes, though.

Comment (Mothblocks) [reply code/game/objects/structures/mirror.dm:23]: You should, one is just for directionals which you'd want to keep, and the other is `magic`.

If you only don't want it on `magic`, you should move all of this out into its own proc, then override that proc in the magic subtype, or disable it through a var.

**Comment by ATH1909** [comment]
I'm still working on this, but I have a project for college I need to finish first.

**Comment by optimumtact** [comment]
open a new PR for this when you're actually ready to complete it (freeze exemption)

**Pull request closed without merging**